src/lib/comms/sol-socket-impl-linux.c

```diff
@@ -33,42 +33,30 @@
 struct sol_socket_linux {
     struct sol_socket base;
 
+    bool (*on_can_read)(void *data, struct sol_socket *s);
+    bool (*on_can_write)(void *data, struct sol_socket *s);
+    struct sol_fd *watch;
+    const void *data;
     int fd;
-    struct {
-        bool (*cb)(void *data, struct sol_socket *s);
-        const void *data;
-        struct sol_fd *watch;
-    } read, write;
 };
 
 static bool
-on_socket_read(void *data, int fd, unsigned int flags)
+on_socket_event(void *data, int fd, unsigned int flags)
 {
     struct sol_socket_linux *s = data;
     bool ret;
 
-    ret = s->read.cb((void *)s->read.data, &s->base);
-    if (!ret) {
-        s->read.cb = NULL;
-        s->read.data = NULL;
-        s->read.watch = NULL;
+    if (flags & SOL_FD_FLAGS_IN) {
+        ret = s->on_can_read((void *)s->data, &s->base);
+        if (!ret)
+            sol_fd_unset_flags(s->watch, SOL_FD_FLAGS_IN);
+    } else if (flags & SOL_FD_FLAGS_OUT) {
+        ret = s->on_can_write((void *)s->data, &s->base);
+        if (!ret)
+            sol_fd_unset_flags(s->watch, SOL_FD_FLAGS_OUT);
     }
-    return ret;
-}
-
-static bool
-on_socket_write(void *data, int fd, unsigned int flags)
-{
-    struct sol_socket_linux *s = data;
-    bool ret;
 
-    ret = s->write.cb((void *)s->write.data, &s->base);
-    if (!ret) {
-        s->write.cb = NULL;
-        s->write.data = NULL;
-        s->write.watch = NULL;
-    }
-    return ret;
+    return true;
```

Comment:
Just a question (I don't know the answer), do we want to keep the watch even if there are no flags set? Something to think about.